### PR TITLE
Feature: support barcode upscaling for better detection of small barcodes

### PR DESCRIPTION
**Pull request by bmachek**
## Proposed change

My QR code labels weren't recognized during ASN barcode detection in paperless, as they seemed to be to small for pyzbar to detect them. I cannot use zxing, as I run paperless on arm64. Couldn't even test zxing, because I currently do not own a x86 device ;-)

I figured that converting the PDF pages with a dpi of 600 and then upscale the resulting image by 1.5 fixes my issue.

## Type of change

I added the ability to paperless to configure an upscale factor and the dpi used in the conversion.
Default values go to 1.0 upscale factor and dpi of 300, so nothing changes for users that are unaffected by this issue. The code won't try to upscale the image by 1.0, it skips the step.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by shamoon** [comment]
Thanks for the PR.

Yea there have been issues with this in the past, eg [#2516](https://github.com/paperless-ngx/paperless-ngx/issues/2516). I don't use the barcode feature, nor have I spent much time with that part of our code. 

In general this looks OK to me but Im always a little hesitant to add more config options, we're sort of at config overload already. Since we dont track anything, we dont have a good way to track how many people want / would use / would benefit from something like this (also a thought to opening discussions before PRs). But I do appreciate the 'sane' defaults and that this would leave things as-is for existing users.

Perhaps other folks will have more thoughts.

One other thing is that we like to include testing of new code, so if you could add that, eg https://github.com/paperless-ngx/paperless-ngx/blob/dev/src/documents/tests/test_barcodes.py

**Comment by stumpylog** [comment]
I'm alright with adding the settings.  Though I do know we're very, very configurable at this point.

**Comment by bmachek** [comment]
Ok, thanks for the feedback.
@shamoon Will add some tests to test_barcodes.py soon.

**Review by stumpylog — CHANGES_REQUESTED** [review]

File: src/documents/barcodes.py [hunk 1]
```diff
@@ -203,11 +203,18 @@ def detect(self) -> None:
...
             pages_from_path = convert_from_path(
                 self.pdf_file,
-                dpi=300,
+                dpi=settings.CONSUMER_ASN_BARCODE_DPI,
                 output_folder=self.temp_dir.name,
             )
 
             for current_page_number, page in enumerate(pages_from_path):
+                factor = settings.CONSUMER_ASN_BARCODE_UPSCALE
+                if factor != 1.0:
```
[the remark this answers is not in the document]
Comment (bmachek) [reply src/documents/barcodes.py]: Scaling by 1.0 didn't was already ruled out by the old code. In the new commit downscaling is also disabled, as I can't think of a use case.

File: docs/configuration.md [hunk 2]
```diff
@@ -1095,6 +1095,22 @@ barcode.
 
     Defaults to "ASN"
 
+`PAPERLESS_CONSUMER_ASN_BARCODE_UPSCALE=<float>`
```
[the remark this answers is not in the document]
Comment (bmachek) [reply docs/configuration.md]: Rename both upscale and dpi variables accordingly.

**Comment by bmachek** [comment]
> One other thing is that we like to include testing of new code, so if you could add that, eg https://github.com/paperless-ngx/paperless-ngx/blob/dev/src/documents/tests/test_barcodes.py

Added a new PDF in samples and wrote a pytest.

**Comment by shamoon** [comment]
Thanks, I would perhaps include the before / after in that test (i.e. that it fails without the upscaling)

**Comment by bmachek** [comment]
> Thanks, I would perhaps include the before / after in that test (i.e. that it fails without the upscaling)

Did that by writing another test, with default settings for dpi and upscale. Maybe this test will fail someday, when pyzbar gets better or zxing becomes default. But IMHO it's OK for now.

File: paperless.conf.example
Comment (shamoon) [inline paperless.conf.example:1]: The new vars should have their default values here

File: src/documents/tests/test_barcodes.py [hunk 3]
```diff
@@ -906,6 +906,45 @@ def test_asn_too_large(self):
                 input_doc,
             )
 
+    def test_scan_file_for_qrcode_without_upscale(self):
+        """
+        ?????
```
Comment (shamoon) [inline src/documents/tests/test_barcodes.py:911]:
```suggestion
```

File: src/documents/tests/test_barcodes.py [hunk 4]
```diff
@@ -906,6 +906,45 @@ def test_asn_too_large(self):
...
+
+        with BarcodeReader(test_file, "application/pdf") as reader:
+            reader.detect()
+            self.assertEqual(len(reader.barcodes), 0)
+
+    @override_settings(CONSUMER_BARCODE_DPI=600)
+    @override_settings(CONSUMER_BARCODE_UPSCALE=1.5)
+    def test_scan_file_for_qrcode_with_upscale(self):
+        """
+        ?????
```
Comment (shamoon) [inline src/documents/tests/test_barcodes.py:931]:
```suggestion
```

**Comment by bmachek** [comment]
I made the requested changes.

The CI runs using ZXing for my pytest methods show that the results differ for negative test without upscaling.
ZXing doesn't seem to need upscaling or higher dpi to detect the small QR code in the test PDF.
This puts the functionality upscaling/dpi to the state of a workaround for the build issues of ZXing on aarch64.
So it maybe removed once ZXing settles its problems there.

**Comment by shamoon** [comment]
Yea thats part of why I wanted the testing, theres a bit of overlapping 'workarounds' here (making both barcode libraries available itself is a sort of workaround).

Thanks, perhaps you can throw in a comment to the tests to explain the above a bit (for some future date). Otherwise, Im good with this, though I will leave final approval to the backend folks (the main person of which is currently AFK so there may be a small delay in finalizing this).

Thanks again

**Review by stumpylog — APPROVED** [review]
I'm good with the changes.  I'll leave it up the @shamoon if this merges to `dev` or if that happens a little later (after other bugfix releases)

**Comment by shamoon** [comment]
Cheers all, thanks for all the work @bmachek

**Pull request merged**